File: src/main/kotlin/io/github/muindor/tcresearchsolver/solver/Board.kt
package io.github.muindor.tcresearchsolver.solver

// ---------------------------------------------------------------------------
// CellState — sealed class (spec §2.2)
// ---------------------------------------------------------------------------

sealed class CellState {
    object Dead : CellState()
    data class Anchor(val aspect: Aspect) : CellState()
    object Empty : CellState()
    data class Placed(val aspect: Aspect, val locked: Boolean) : CellState()
}

// ---------------------------------------------------------------------------
// Board — radius + LinkedHashMap<String, CellState> (absent key == Empty)
// ---------------------------------------------------------------------------

class Board(val radius: Int, val cells: LinkedHashMap<String, CellState>)

fun createBoard(radius: Int): Board {
    if (radius < 2 || radius > 5) throw IllegalArgumentException("radius must be an integer 2..5, got $radius")
    return Board(radius, LinkedHashMap())
}

fun getState(board: Board, h: Hex): CellState {
    if (!isOnBoard(h, board.radius)) throw IllegalArgumentException("hex ${hexKey(h)} is off board (R=${board.radius})")
    return board.cells[hexKey(h)] ?: CellState.Empty
}

fun setState(board: Board, h: Hex, s: CellState) {
    if (!isOnBoard(h, board.radius)) throw IllegalArgumentException("hex ${hexKey(h)} is off board (R=${board.radius})")
    val key = hexKey(h)
    if (s is CellState.Empty) board.cells.remove(key)
    else board.cells[key] = s
}

// ---------------------------------------------------------------------------
// FilledCell — analogous to the TS interface
// ---------------------------------------------------------------------------

data class FilledCell(val hex: Hex, val aspect: Aspect, val locked: Boolean, val isAnchor: Boolean)

fun filledCells(board: Board): List<FilledCell> {
    val out = ArrayList<FilledCell>()
    for (h in boardCells(board.radius)) {
        val s = getState(board, h)
        when (s) {
            is CellState.Anchor -> out.add(FilledCell(h, s.aspect, false, true))
            is CellState.Placed -> out.add(FilledCell(h, s.aspect, s.locked, false))
            else -> {}
        }
    }
    return out
}

fun anchorCells(board: Board): List<Pair<Hex, Aspect>> =
    filledCells(board).filter { it.isAnchor }.map { Pair(it.hex, it.aspect) }

/** On-board neighbors that are filled (ANCHOR or PLACED). */
fun filledNeighbors(board: Board, h: Hex): List<FilledCell> {
    val out = ArrayList<FilledCell>()
    for (n in neighborsOf(h)) {
        if (!isOnBoard(n, board.radius)) continue
        val s = getState(board, n)
        when (s) {
            is CellState.Anchor -> out.add(FilledCell(n, s.aspect, false, true))
            is CellState.Placed -> out.add(FilledCell(n, s.aspect, s.locked, false))
            else -> {}
        }
    }
    return out
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

enum class ValidationErrorType {
    INVALID_LINK, SAME_ASPECT_ADJACENT, ANCHORS_DISCONNECTED, PLACED_ON_DEAD, MALFORMED
}

data class ValidationError(val type: ValidationErrorType, val cells: List<Hex>)
data class ValidationResult(val valid: Boolean, val errors: List<ValidationError>)

fun validate(data: AspectData, board: Board): ValidationResult {
    val errors = ArrayList<ValidationError>()
    val filled = filledCells(board)
    val filledKeys = LinkedHashSet(filled.map { hexKey(it.hex) })
    val aspectAt = LinkedHashMap<String, Aspect>().apply { for (c in filled) put(hexKey(c.hex), c.aspect) }

    // 1) pairwise adjacency validity (each undirected pair once)
    for (c in filled) {
        for (n in neighborsOf(c.hex)) {
            val nk = hexKey(n)
            if (!filledKeys.contains(nk)) continue
            if (hexKey(c.hex) >= nk) continue  // dedupe ordered pair (G6)
            val a = c.aspect
            val b = aspectAt[nk]!!
            if (a == b) errors.add(ValidationError(ValidationErrorType.SAME_ASPECT_ADJACENT, listOf(c.hex, n)))
            else if (!isValidLink(data, a, b)) errors.add(ValidationError(ValidationErrorType.INVALID_LINK, listOf(c.hex, n)))
        }
    }

    // 2) anchors connectivity (>=2 anchors must share one filled component)
    if (!anchorsConnectedInternal(board)) {
        errors.add(ValidationError(ValidationErrorType.ANCHORS_DISCONNECTED, anchorCells(board).map { it.first }))
    }

    return ValidationResult(errors.isEmpty(), errors)
}

private fun anchorsConnectedInternal(board: Board): Boolean {
    val anchors = anchorCells(board)
    if (anchors.size <= 1) return true
    val filled = LinkedHashSet(filledCells(board).map { hexKey(it.hex) })
    // BFS from first anchor over filled adjacency
    val start = hexKey(anchors[0].first)
    val seen = LinkedHashSet<String>()
    seen.add(start)
    val queue = ArrayDeque<Hex>()
    queue.add(anchors[0].first)
    while (queue.isNotEmpty()) {
        val cur = queue.removeLast()
        for (n in neighborsOf(cur)) {
            val nk = hexKey(n)
            if (filled.contains(nk) && seen.add(nk)) {
                queue.add(n)
            }
        }
    }
    return anchors.all { seen.contains(hexKey(it.first)) }
}

fun allAnchorsConnected(board: Board): Boolean = anchorsConnectedInternal(board)

/** A finished solution: fully valid AND all anchors in one component. */
fun isComplete(data: AspectData, board: Board): Boolean = validate(data, board).valid

// ---------------------------------------------------------------------------
// Serialization
// ---------------------------------------------------------------------------

const val BOARD_SCHEMA_VERSION = 1

data class SerializedCell(val coord: String, val state: String, val aspect: Aspect? = null, val locked: Boolean? = null)
data class SerializedBoard(val schemaVersion: Int, val radius: Int, val cells: List<SerializedCell>)

fun serializeBoard(board: Board): SerializedBoard {
    val cells = ArrayList<SerializedCell>()
    for ((key, s) in board.cells) {
        when (s) {
            is CellState.Empty -> continue
            is CellState.Dead -> cells.add(SerializedCell(key, "DEAD"))
            is CellState.Anchor -> cells.add(SerializedCell(key, "ANCHOR", s.aspect))
            is CellState.Placed -> cells.add(SerializedCell(key, "PLACED", s.aspect, s.locked))
        }
    }
    return SerializedBoard(BOARD_SCHEMA_VERSION, board.radius, cells)
}

/** Overload: accept a typed DTO directly, mirroring the TS round-trip where
 *  serializeBoard's plain-object output is passable to deserializeBoard. */
fun deserializeBoard(data: AspectData, dto: SerializedBoard): Board {
    if (dto.schemaVersion != BOARD_SCHEMA_VERSION)
        throw IllegalArgumentException("board: unsupported schemaVersion ${dto.schemaVersion}")
    val board = createBoard(dto.radius)
    for (c in dto.cells) {
        val hex = parseHexKey(c.coord)
        if (!isOnBoard(hex, board.radius)) throw IllegalArgumentException("board: coord ${c.coord} off radius ${board.radius}")
        fun checkAspect(a: Aspect?): Aspect {
            if (a == null || !data.universe.contains(a)) throw IllegalArgumentException("board: unknown aspect '${a}'")
            return a
        }
        when (c.state) {
            "DEAD" -> setState(board, hex, CellState.Dead)
            "ANCHOR" -> setState(board, hex, CellState.Anchor(checkAspect(c.aspect)))
            "PLACED" -> setState(board, hex, CellState.Placed(checkAspect(c.aspect), c.locked == true))
            else -> throw IllegalArgumentException("board: bad cell.state '${c.state}'")
        }
    }
    return board
}

fun deserializeBoard(data: AspectData, raw: Any?): Board {
    if (raw is SerializedBoard) return deserializeBoard(data, raw)
    if (raw == null || raw !is Map<*, *>) throw IllegalArgumentException("board: not an object")
    val obj = raw as Map<*, *>
    val radiusRaw = obj["radius"]
    val radius = when (radiusRaw) {
        is Int -> radiusRaw
        is Long -> radiusRaw.toInt()
        is Double -> if (radiusRaw == radiusRaw.toLong().toDouble()) radiusRaw.toInt() else throw IllegalArgumentException("board: bad radius $radiusRaw")
        else -> throw IllegalArgumentException("board: bad radius $radiusRaw")
    }
    if (radius < 2 || radius > 5) throw IllegalArgumentException("board: bad radius $radius")

    val ver = obj["schemaVersion"]
    val verInt = when (ver) {
        is Int -> ver
        is Long -> ver.toInt()
        is Double -> if (ver == ver.toLong().toDouble()) ver.toInt() else throw IllegalArgumentException("board: unsupported schemaVersion $ver")
        else -> throw IllegalArgumentException("board: unsupported schemaVersion $ver")
    }
    if (verInt != BOARD_SCHEMA_VERSION) throw IllegalArgumentException("board: unsupported schemaVersion $verInt")

    val cellsRaw = obj["cells"] ?: throw IllegalArgumentException("board: cells must be an array")
    if (cellsRaw !is List<*>) throw IllegalArgumentException("board: cells must be an array")

    val board = createBoard(radius)
    for (cellRaw in cellsRaw) {
        if (cellRaw == null || cellRaw !is Map<*, *>) throw IllegalArgumentException("board: bad cell")
        val cell = cellRaw as Map<*, *>
        val coordRaw = cell["coord"]
        if (coordRaw !is String) throw IllegalArgumentException("board: cell.coord must be a string")
        val hex = parseHexKey(coordRaw)
        if (!isOnBoard(hex, board.radius)) throw IllegalArgumentException("board: coord $coordRaw off radius ${board.radius}")

        fun checkAspect(a: Any?): Aspect {
            if (a !is String || !data.universe.contains(a)) throw IllegalArgumentException("board: unknown aspect '${a}'")
            return a
        }

        when (val stateRaw = cell["state"]) {
            "DEAD" -> setState(board, hex, CellState.Dead)
            "ANCHOR" -> setState(board, hex, CellState.Anchor(checkAspect(cell["aspect"])))
            "PLACED" -> setState(board, hex, CellState.Placed(checkAspect(cell["aspect"]), cell["locked"] == true))
            else -> throw IllegalArgumentException("board: bad cell.state '$stateRaw'")
        }
    }
    return board
}

File: reference/ts-solver/core/board.ts
import type { Aspect, AspectData } from '../data/aspects';
import { type Hex, hexKey, parseHexKey, isOnBoard, boardCells, neighborsOf } from './hex';
import { isValidLink } from './aspectGraph';

export type CellState =
  | { kind: 'DEAD' }
  | { kind: 'ANCHOR'; aspect: Aspect }
  | { kind: 'EMPTY' }
  | { kind: 'PLACED'; aspect: Aspect; locked: boolean };

export interface Board {
  readonly radius: number;
  /** Only non-EMPTY cells are stored; absent on-board key => EMPTY. */
  readonly cells: Map<string, CellState>;
}

export function createBoard(radius: number): Board {
  if (!Number.isInteger(radius) || radius < 2 || radius > 5) {
    throw new Error(`radius must be an integer 2..5, got ${radius}`);
  }
  return { radius, cells: new Map() };
}

export function getState(board: Board, h: Hex): CellState {
  if (!isOnBoard(h, board.radius)) throw new Error(`hex ${hexKey(h)} is off board (R=${board.radius})`);
  return board.cells.get(hexKey(h)) ?? { kind: 'EMPTY' };
}

export function setState(board: Board, h: Hex, s: CellState): void {
  if (!isOnBoard(h, board.radius)) throw new Error(`hex ${hexKey(h)} is off board (R=${board.radius})`);
  if (s.kind === 'EMPTY') board.cells.delete(hexKey(h));
  else board.cells.set(hexKey(h), s);
}

export interface FilledCell { hex: Hex; aspect: Aspect; locked: boolean; isAnchor: boolean; }

export function filledCells(board: Board): FilledCell[] {
  const out: FilledCell[] = [];
  for (const h of boardCells(board.radius)) {
    const s = getState(board, h);
    if (s.kind === 'ANCHOR') out.push({ hex: h, aspect: s.aspect, locked: false, isAnchor: true });
    else if (s.kind === 'PLACED') out.push({ hex: h, aspect: s.aspect, locked: s.locked, isAnchor: false });
  }
  return out;
}

export function anchorCells(board: Board): Array<{ hex: Hex; aspect: Aspect }> {
  return filledCells(board).filter((c) => c.isAnchor).map((c) => ({ hex: c.hex, aspect: c.aspect }));
}

/** On-board neighbors that are filled (ANCHOR or PLACED). */
export function filledNeighbors(board: Board, h: Hex): FilledCell[] {
  const out: FilledCell[] = [];
  for (const n of neighborsOf(h)) {
    if (!isOnBoard(n, board.radius)) continue;
    const s = getState(board, n);
    if (s.kind === 'ANCHOR') out.push({ hex: n, aspect: s.aspect, locked: false, isAnchor: true });
    else if (s.kind === 'PLACED') out.push({ hex: n, aspect: s.aspect, locked: s.locked, isAnchor: false });
  }
  return out;
}

export type ValidationErrorType =
  | 'INVALID_LINK' | 'SAME_ASPECT_ADJACENT' | 'ANCHORS_DISCONNECTED' | 'PLACED_ON_DEAD' | 'MALFORMED';

export interface ValidationError { type: ValidationErrorType; cells: Hex[]; }
export interface ValidationResult { valid: boolean; errors: ValidationError[]; }

export function validate(data: AspectData, board: Board): ValidationResult {
  const errors: ValidationError[] = [];
  const filled = filledCells(board);
  const filledKeys = new Set(filled.map((c) => hexKey(c.hex)));
  const aspectAt = new Map(filled.map((c) => [hexKey(c.hex), c.aspect]));

  // 1) pairwise adjacency validity (each undirected pair once)
  for (const c of filled) {
    for (const n of neighborsOf(c.hex)) {
      const nk = hexKey(n);
      if (!filledKeys.has(nk)) continue;
      if (hexKey(c.hex) >= nk) continue; // dedupe ordered pair
      const a = c.aspect;
      const b = aspectAt.get(nk)!;
      if (a === b) errors.push({ type: 'SAME_ASPECT_ADJACENT', cells: [c.hex, n] });
      else if (!isValidLink(data, a, b)) errors.push({ type: 'INVALID_LINK', cells: [c.hex, n] });
    }
  }

  // 2) anchors connectivity (>=2 anchors must share one filled component)
  if (!anchorsConnectedInternal(board)) {
    errors.push({ type: 'ANCHORS_DISCONNECTED', cells: anchorCells(board).map((a) => a.hex) });
  }

  return { valid: errors.length === 0, errors };
}

function anchorsConnectedInternal(board: Board): boolean {
  const anchors = anchorCells(board);
  if (anchors.length <= 1) return true;
  const filled = new Set(filledCells(board).map((c) => hexKey(c.hex)));
  // BFS from first anchor over filled adjacency
  const start = hexKey(anchors[0]!.hex);
  const seen = new Set<string>([start]);
  const queue = [anchors[0]!.hex];
  while (queue.length) {
    const cur = queue.pop()!;
    for (const n of neighborsOf(cur)) {
      const nk = hexKey(n);
      if (filled.has(nk) && !seen.has(nk)) {
        seen.add(nk);
        queue.push(n);
      }
    }
  }
  return anchors.every((a) => seen.has(hexKey(a.hex)));
}

export function allAnchorsConnected(board: Board): boolean {
  return anchorsConnectedInternal(board);
}

/** A finished solution: fully valid AND all anchors in one component. */
export function isComplete(data: AspectData, board: Board): boolean {
  return validate(data, board).valid;
}

export const BOARD_SCHEMA_VERSION = 1;

export interface SerializedCell { coord: string; state: 'DEAD' | 'ANCHOR' | 'PLACED'; aspect?: Aspect; locked?: boolean; }
export interface SerializedBoard { schemaVersion: number; radius: number; cells: SerializedCell[]; }

export function serializeBoard(board: Board): SerializedBoard {
  const cells: SerializedCell[] = [];
  for (const [key, s] of board.cells) {
    if (s.kind === 'EMPTY') continue;
    if (s.kind === 'DEAD') cells.push({ coord: key, state: 'DEAD' });
    else if (s.kind === 'ANCHOR') cells.push({ coord: key, state: 'ANCHOR', aspect: s.aspect });
    else cells.push({ coord: key, state: 'PLACED', aspect: s.aspect, locked: s.locked });
  }
  return { schemaVersion: BOARD_SCHEMA_VERSION, radius: board.radius, cells };
}

export function deserializeBoard(data: AspectData, raw: unknown): Board {
  if (typeof raw !== 'object' || raw === null) throw new Error('board: not an object');
  const obj = raw as Record<string, unknown>;
  const radius = obj.radius;
  if (!Number.isInteger(radius) || (radius as number) < 2 || (radius as number) > 5) {
    throw new Error(`board: bad radius ${String(radius)}`);
  }
  // schemaVersion migration hook (only v1 exists today).
  const ver = obj.schemaVersion;
  if (ver !== BOARD_SCHEMA_VERSION) throw new Error(`board: unsupported schemaVersion ${String(ver)}`);
  if (!Array.isArray(obj.cells)) throw new Error('board: cells must be an array');

  const board = createBoard(radius as number);
  for (const c of obj.cells as unknown[]) {
    if (typeof c !== 'object' || c === null) throw new Error('board: bad cell');
    const cell = c as Record<string, unknown>;
    if (typeof cell.coord !== 'string') throw new Error('board: cell.coord must be a string');
    const hex = parseHexKey(cell.coord);
    if (!isOnBoard(hex, board.radius)) throw new Error(`board: coord ${cell.coord} off radius ${board.radius}`);
    const checkAspect = (a: unknown): Aspect => {
      if (typeof a !== 'string' || !data.universe.has(a)) throw new Error(`board: unknown aspect '${String(a)}'`);
      return a;
    };
    switch (cell.state) {
      case 'DEAD': setState(board, hex, { kind: 'DEAD' }); break;
      case 'ANCHOR': setState(board, hex, { kind: 'ANCHOR', aspect: checkAspect(cell.aspect) }); break;
      case 'PLACED': setState(board, hex, { kind: 'PLACED', aspect: checkAspect(cell.aspect), locked: cell.locked === true }); break;
      default: throw new Error(`board: bad cell.state '${String(cell.state)}'`);
    }
  }
  return board;
}
